crates/hir_ty/src/infer/expr.rs

```diff
@@ -430,11 +430,33 @@ impl<'a> InferenceContext<'a> {
                         name.as_tuple_index().and_then(|idx| substs.0.get(idx).cloned())
                     }
                     Ty::Adt(AdtId(hir_def::AdtId::StructId(s)), parameters) => {
-                        self.db.struct_data(s).variant_data.field(name).map(|local_id| {
-                            let field = FieldId { parent: s.into(), local_id };
-                            self.write_field_resolution(tgt_expr, field);
-                            self.db.field_types(s.into())[field.local_id].clone().subst(&parameters)
-                        })
+                        let def_db = self.db.upcast();
+                        let adt_module = hir_def::AdtId::from(s).module(def_db);
+                        let adt_resolver = adt_module.resolver(def_db);
```

Comment:
I think we should avoid resolving the visibilities every time here, and create a query for the resolved field visibilities (somewhat similar to the `field_types` query) instead.

Reply:
Okay. I'll try that in the next revision.